solutions/cf-920E.mdx

```diff
@@ -0,0 +1,236 @@
+
+
+[Official Editorial](https://codeforces.com/blog/entry/57516)
+
+
+## Explanation
+## Solution 1 - Faster DFS
+
+Doing a standard DFS would be too slow. To speed it up, we have to do two
+optimizations.
+
+### Optimization 1
+
+You can store whether there is no edge between two nodes using an array of sets, instead of a adjacency list. If the input tells you that nodes $a$ and $b$ do node have an edge between them, you would insert node $a$ at $array[b]$, and node $b$ and $array[a]$ 
+This way you can find if there is no edge between two nodes in
+$\mathcal{O}(\log N)$ time. Using an adjacency matrix would take too much
+memory.
+
+### Optimization 2
+
+Rather than using a standard boolean array to check if a node is not visited, you
+can use a set of unvisited nodes. This way you only loop through the nodes that
+have not been visited yet and you can find the next unvisited node in $\mathcal{O}(\log N)$. You need to use ```upper_bound``` to find
+the next unvisited node after a DFS - an  iterator may be invalidated due to the
+node it is pointing to being destoryed when you remove it from the unvisited set.
+
+### Proof
+
+We know that a node can either be skipped or visited. We also know that once a
+node is visited it cannot be revisited. Therefore you will always visit $N$ nodes once. We also know that the sum of skips can be at most $2M$ (each        
+non-edge skips two nodes). Therefore the overall complexity will be
+$O(N\log N+M\log N)$.
+
+<LanguageSection>
+<CPPSection>
+
+## Implementation
+**Time Complexity:** $\mathcal{O}(N\log N+M\log N)$
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+set<int> adj[200000];
+//the set of nodes that have not been visited
+set<int> unvis;
+// sz[i] = size of the ith connected component
+int sz[200000];
+//current amount of groups
+int cur = 0;
+void dfs(int x){
+	sz[cur]++;
+	auto it = unvis.begin();
+	while(it!=unvis.end()){
+		//if there is no edge between x and *it, skip
+		if(adj[x].count(*it)){
+			it++;
+		}else{
+			//there is a edge between last and x
+			int last = *it;
+			//note it = unvis.erase(it) doenst work here because it could be erased later
+			unvis.erase(it);
+			dfs(last);
+			//find the node after last
+			it = unvis.upper_bound(last);
+		}
+	}
+}
+int main(){
+	int n, m;
+	cin>>n>>m;
+	for(int i = 0; i<m; i++){
+		int a, b;
+		cin>>a>>b;
+		a--;
+		b--;
+		adj[a].insert(b);
+		adj[b].insert(a);
+	}
+	for(int i = 0; i<n; i++){
+		unvis.insert(i);
+	}
+	for(int i = 0; i<n; i++){
+		auto it = unvis.find(i);
+		if(it!=unvis.end()){
+			unvis.erase(it);
+			dfs(i);
+			cur++;
+		}
+	}
+	cout<<cur<<endl;
+	sort(sz, sz+cur);
+	for(int i = 0; i<cur; i++){
+		cout<<sz[i]<<' ';
+	}
+	cout<<endl;
+}
+```
+</CPPSection>
+</LanguageSection>
+
+## Solution 2 - Selective DFS
+
+Let $N =$ number of nodes in the graph
+
+$n =$ an abitrary node 
+
+$\texttt{degree}(n)=$ the number of neighbooring nodes (number of different nodes $n$ is directly connected to). 
+
+<br></br>
```

Comment:
What's this

Reply:
I wanted to add a bigger gap between the terminology and the solution. <br> is break in html, it creates a line break in text.

Reply:
The header generally provides enough spacing, so I don't think you need it